File: examples/missing/example_schafer1.R
library("MASS")
x <- mmstat4::ghload("gs50.rds")
plot(x[,1:2])
df <- data.frame(V1=x$x, V2=x$ymar)  # y, ymcar, ymar, ymnar
lm <- lm(V2~V1, data=df)
b  <- coefficients(lm)
na_impute <- "preddist"   # delete, mean subst., hot deck, cond. mean, pred. dist.
na        <- is.na(df[,2])
df[na,2]  <- switch(na_impute, rep(NA, sum(na)),
               mean=rep(mean(x[!na,2]), sum(na)),
               hotdeck=sample(x[!na,2], sum(na), replace=TRUE),
               condmean=b[1]+b[2]*x[na,1],
               preddist=b[1]+b[2]*x[na,1]+
               sample(residuals(lm), sum(na), replace=TRUE))
points(df, col="red", pch=19, cex=0.75)

File: examples/missing/example_schafer1.py
import numpy as np, pandas as pd, matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
x  = r["mmstat4::ghload"]("gs50.rds")
df = pd.DataFrame({'V1': x['x'], 'V2': x['ymar']}) 
na = df['V2'].isna()
plt.figure(); plt.plot(x['x'], x['y'], "o", markersize=5)
lm = LinearRegression(); lm.fit(df.loc[~na,['V1']], df.loc[~na,'V2'])
lp = lm.predict(df.loc[~na, ['V1']])
residuals = df.loc[~na, 'V2'] - lp
sw_method = {
    "delete": df.dropna(),  # Actual removal of missing rows
    "mean": df.loc[~na, 'V2'].mean(),
    "hotdeck": np.random.choice(df.loc[~na, 'V2'], size=na.sum(), 
                                replace=True),
    "condmean": lm.predict(df.loc[na, ['V1']]),
    "preddist": condmean_values + np.random.choice(residuals, 
                size=na.sum(), replace=True)
}
df.loc[na, 'V2'] = sw_method["mean"]
plt.plot(df['V1'], df['V2'], 'ro', markersize=3); plt.show()
